src/cosalib/ostree.py

```diff
@@ -0,0 +1,221 @@
+                       ):
+    """
+    Get the size of the OSTree.
+    """
+    log.debug(f"Estimating tree size of {ref} at {repo}")
+    r = OSTree.Repo.new(Gio.File.new_for_path(repo))
+    r.open(None)
+
+    [_, rev] = r.resolve_rev(ref, False)
+    [_, reachable] = r.traverse_commit(rev, 0, None)
+    n_meta = 0
+    blks_meta = 0
+    n_regfiles = 0
+    blks_regfiles = 0
+    n_symlinks = 0
+    blks_symlinks = 0
+    for k, v in reachable.items():
+        csum, objtype = k.unpack()
+        if objtype == OSTree.ObjectType.FILE:
+            [_, _, finfo, _] = r.load_file(csum, None)
+            if finfo.get_file_type() == Gio.FileType.REGULAR:
+                n_regfiles += 1
+                sz = finfo.get_size()
+                blks_regfiles += (sz // blksize) + 1
+            else:
+                n_symlinks += 1
+                sz = len(finfo.get_symlink_target())
+                blks_symlinks += (sz // blksize) + 1
+        else:
+            [_, sz] = r.query_object_storage_size(objtype, csum, None)
+            n_meta += 1
+            blks_meta += (sz // blksize) + 1
+
+    mb = 1024 * 1024
+    blks_per_mb = mb // blksize
+    total_data_mb = (
+        blks_meta + blks_regfiles + blks_symlinks) // blks_per_mb
+    n_inodes = n_meta + n_regfiles + n_symlinks
+    total_inode_mb = 1 + ((n_inodes * isize) // mb)
+    total_mb = total_data_mb + total_inode_mb
+    add_percent = metadata_overhead_percent + add_percent
+    add_percent_modifier = (100.0 + add_percent) / 100.0
+    estimate_mb = int(total_mb * add_percent_modifier) + 1
+    return {
+        'meta': {
+            'count': n_meta,
+            'blocks': blks_meta,
+        },
+        'regfiles': {
+            'count': n_regfiles,
+            'blocks': blks_regfiles},
+        'symlinks': {
+            'count': n_symlinks,
+            'blocks': blks_symlinks
+        },
+        'inodes': {
+            'count': n_inodes,
+            'mb': total_inode_mb,
+        },
+        'estimate-mb': {
+            'base': total_mb,
+            'final': estimate_mb
+        },
+    }
+
+
+class OSTreeException(Exception):
+    pass
+
+
+class _BuildOSTree():
+    """
+    BuildOSTree interacts with OS Trees written to <dir>/tmp/repo from the
+    CoreOS Assembler command "cosa build."
+
+    The core consumer is cosalib.metal.MetalVariant.
+    """
+    def __init__(self, build):
+        self._build = build
+        if not isinstance(build, _Build):
+            raise OSTreeException("BuildOStree expects a _Build Object")
+
+        self._ostree_repo = os.path.join(build.workdir, "tmp", "repo")
+        if not os.path.exists(self._ostree_repo):
+            raise OSTreeException(
+                f"missing ostree repo at {self._ostree_repo}")
+
+        # find the compose-*json file to discern the ref we expect
+        try:
+            os.chdir(os.path.join(build.workdir, "tmp"))
```

Comment:
Using `chdir()` is very dangerous in multi-threaded processes - avoiding it is another thing Java got right a long time ago.